Add support for external version numbers

vvcephei opened the pull request:
I'm just pushing an initial commit so I can see the tests run in TravisCI.

For some reason, I'm unable to run the tests locally. They fail with this error:

```
$ bundle exec rspec
The signal EXIT is in use by the JVM and will not work correctly on this platform
bundler: failed to load command: rspec (/home/john/repos/baz/logstash-output-elasticsearch/vendor/bundle/jruby/1.9/bin/rspec)
RuntimeError: 

        you might need to reinstall the gem which depends on the missing jar or in case there is Jars.lock then resolve the jars with `lock_jars` command

no such file to load -- org/apache/logging/log4j/log4j-core/2.6.2/log4j-core-2.6.2 (LoadError)
... (stacktrace) ...
```

vvcephei commented:
This PR is related to #405 #365 #279.

untergeek commented:
I can see that this is a work in progress.  Can you let us know when you want us to review it?

Also, with so many changes which are only there for debugging, when you're ready, will you please squash all of the commits to a single commit?

vvcephei commented:
@untergeek, thanks for the note. Yep, I will squash it once I'm happy and then ask for a review.

Hopefully, you don't mind the noise. I tried for a couple of hours to get the tests running locally to no avail, so I'm offloading the job to your TravisCI PR builder.

vvcephei commented:
for posterity, the resolution to my problem with log4j was that I just didn't have the artifacts in my ~/.m2/repository. I'm not sure if bundler is supposed to fetch them or not, or why no one else reported this problem. Perhaps everyone working on logstash has happened to build elasticsearch recently and pulled in those dependencies that way?

Any way, I was able to get it working my just installing these in my local mvn repo:
```
      <dependency>
        <groupId>org.apache.logging.log4j</groupId>
        <artifactId>log4j-core</artifactId>
        <version>2.6.2</version>
      </dependency>
      <dependency>
        <groupId>org.apache.logging.log4j</groupId>
        <artifactId>log4j-1.2-api</artifactId>
        <version>2.6.2</version>
      </dependency>
      <dependency>
        <groupId>com.fasterxml.jackson.core</groupId>
        <artifactId>jackson-core</artifactId>
        <version>2.7.4</version>
      </dependency>
      <dependency>
        <groupId>com.fasterxml.jackson.core</groupId>
        <artifactId>jackson-databind</artifactId>
        <version>2.7.4</version>
      </dependency>
      <dependency>
        <groupId>com.fasterxml.jackson.dataformat</groupId>
        <artifactId>jackson-dataformat-cbor</artifactId>
        <version>2.7.3</version>
      </dependency>
```

vvcephei commented:
@untergeek Ok, thanks for your patience. The PR is now ready for review.

vvcephei commented:
@jordansissel If you are following this issue, this PR addresses #405 .

vvcephei commented:
No. 1 concern: I'm on the fence about specifying both the version and version_type=external in one config.

On one hand, it's nice not to have too many knobs to turn, especially since some configuration combinations are invalid.

On the other, it seems more sane just to mirror ES's api. Particularly, this would allow you to choose between "external"/"external_gt" and "external_gte", and of course whatever else may be offered in the future.

There's also a trade-off to be made about whether to validate at configuration time or let ES validate the requests when they are sent. Validating the configuration potentially surfaces issues faster and better, but not validating and letting ES do it offers better forward compatibility (and simpler code in the plugin).

In all these decisions, I'm happy to defer to the maintainers. I just made a decision and went with it to get the PR rolling.

andrewvc commented on .gitignore:
```diff
@@ -1,6 +1,12 @@
 .idea
 *~
 .ruby-version
+elasticsearch.tar.gz
```
Thanks for adding these

andrewvc commented on lib/logstash/outputs/elasticsearch/common.rb:
```diff
@@ -123,6 +124,9 @@ def submit(actions)
         if SUCCESS_CODES.include?(status)
           next
+	elsif CONFLICT_CODE == status && action[1][:version_type] == "external"
+	  # When using external versioning, conflicts are basically successes.
```
I'd like to see `action[1]` pulled out into a more semantically named variable. Something like `action_metadata`. Maybe you have a better idea? I don't love that name.

andrewvc commented on lib/logstash/outputs/elasticsearch/common.rb:
```diff
@@ -211,6 +220,7 @@ def safe_bulk(actions)
         sleep_interval = next_sleep_interval(sleep_interval)
         retry unless @stopping.true?
       rescue ::LogStash::Outputs::ElasticSearch::HttpClient::Pool::BadResponseCodeError => e
+        safe_url = ::LogStash::Outputs::ElasticSearch::SafeURL.without_credentials(e.url)
```
Can you rebase this on master? Luckily we no longer need this as of today's merge of #450, `e.url` is now a `LogStash::Util::SafeURI`. For extra safety you can now log `e.url.sanitize`.

andrewvc replied:
This line should pretty much not be here.

andrewvc commented on lib/logstash/outputs/elasticsearch/http_client_builder.rb:
```diff
@@ -44,6 +44,16 @@ def self.build(logger, hosts, params)
+      raise(
+        LogStash::ConfigurationError,
+        "External versioning is not supported by the create action."
+      ) if params['action'] == 'create' and params.fetch('external_version', '') != ''
```
Instead of `!= ''` could you use `String#empty?`.

andrewvc commented on lib/logstash/outputs/elasticsearch/http_client_builder.rb:
```diff
@@ -44,6 +44,16 @@ def self.build(logger, hosts, params)
+      raise(
+        LogStash::ConfigurationError,
+        "External versioning is not supported by the update action. See https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-update.html."
+      ) if params['action'] == 'update' and params.fetch('external_version', '') != ''
```
Same deal with `String#empty?`

andrewvc commented on logstash-output-elasticsearch.gemspec:
```diff
@@ -25,6 +25,7 @@ Gem::Specification.new do |s|
   s.add_development_dependency 'ftw', '~> 0.0.42'
   s.add_development_dependency 'addressable', "~> 2.3.0" # used by FTW. V 2.5.0 is ruby 2.0 only.
   s.add_development_dependency 'logstash-codec-plain'
+  s.add_development_dependency 'json' # used by spec/unit/outputs/elasticsearch/http_client/pool_spec.rb
```
Why do we need this now? It isn't needed in master.

vvcephei replied:
I ... am not sure. It's clearly referenced here: https://github.com/logstash-plugins/logstash-output-elasticsearch/blob/master/spec/unit/outputs/elasticsearch/http_client/pool_spec.rb#L3

Maybe everyone who already works on the project happens to have it on their classpath for other reasons? I figured I ran into it because I started from a blank-slate environment.

andrewvc replied:
hmmmm, interesting. makes sense! This should really just be a plain dependency, not a development_dependency then. Excellent catch :)

vvcephei replied:
Once again, forgive my noobishness, but I think gemspec allows just `add_development_dependency` and `add_runtime_dependency`. If it's only required by tests, shouldn't it be a development dep? Otherwise, it would leak json into any downstream projects.

While I was at it, i thought it would be nifty to make sure json is actually required. I applied this diff:
```diff
diff --git a/logstash-output-elasticsearch.gemspec b/logstash-output-elasticsearch.gemspec
index 5c045c8..6f0b38b 100644
--- a/logstash-output-elasticsearch.gemspec
+++ b/logstash-output-elasticsearch.gemspec
@@ -25,7 +25,6 @@ Gem::Specification.new do |s|
   s.add_development_dependency 'ftw', '~> 0.0.42'
   s.add_development_dependency 'addressable', "~> 2.3.0" # used by FTW. V 2.5.0 is ruby 2.0 only.
   s.add_development_dependency 'logstash-codec-plain'
-  s.add_development_dependency 'json' # used by spec/unit/outputs/elasticsearch/http_client/pool_spec.rb
 
   if RUBY_PLATFORM == 'java'
     s.platform = RUBY_PLATFORM
diff --git a/spec/integration/outputs/sniffer_spec.rb b/spec/integration/outputs/sniffer_spec.rb
index f3d5685..c54591b 100644
--- a/spec/integration/outputs/sniffer_spec.rb
+++ b/spec/integration/outputs/sniffer_spec.rb
@@ -1,7 +1,6 @@
 require "logstash/devutils/rspec/spec_helper"
 require_relative "../../../spec/es_spec_helper"
 require "logstash/outputs/elasticsearch/http_client"
-require "json"
 
 describe "pool sniffer", :integration => true do
   let(:logger) { Cabin::Channel.get }
diff --git a/spec/unit/outputs/elasticsearch/http_client/pool_spec.rb b/spec/unit/outputs/elasticsearch/http_client/pool_spec.rb
index 3406623..5bfe2e6 100644
--- a/spec/unit/outputs/elasticsearch/http_client/pool_spec.rb
+++ b/spec/unit/outputs/elasticsearch/http_client/pool_spec.rb
@@ -1,6 +1,5 @@
 require "logstash/devutils/rspec/spec_helper"
 require "logstash/outputs/elasticsearch/http_client"
-require "json"
 
 describe LogStash::Outputs::ElasticSearch::HttpClient::Pool do
   let(:logger) { Cabin::Channel.get }
diff --git a/spec/unit/outputs/elasticsearch/http_client_spec.rb b/spec/unit/outputs/elasticsearch/http_client_spec.rb
index 3ade145..020b31e 100644
--- a/spec/unit/outputs/elasticsearch/http_client_spec.rb
+++ b/spec/unit/outputs/elasticsearch/http_client_spec.rb
@@ -125,7 +125,6 @@ describe LogStash::Outputs::ElasticSearch::HttpClient do
     subject { described_class.new(base_options) }
 
     context "when items key is available" do
-      require "json"
       let(:bulk_response) {
         LogStash::Json.load ('[{
           "items": [{
@@ -149,7 +148,6 @@ describe LogStash::Outputs::ElasticSearch::HttpClient do
     end
 
     context "when items key is not available" do
-      require "json"
       let(:bulk_response) {
         JSON.parse ('[{
           "took": 4,
diff --git a/spec/unit/outputs/elasticsearch/template_manager_spec.rb b/spec/unit/outputs/elasticsearch/template_manager_spec.rb
index d63fb05..ba8825c 100644
--- a/spec/unit/outputs/elasticsearch/template_manager_spec.rb
+++ b/spec/unit/outputs/elasticsearch/template_manager_spec.rb
@@ -1,7 +1,6 @@
 require "logstash/devutils/rspec/spec_helper"
 require "logstash/outputs/elasticsearch/http_client"
 require "java"
-require "json"
 
 describe LogStash::Outputs::ElasticSearch::TemplateManager do
```

And running the (unit) tests gives the following failures:

```
Failures:

  1) LogStash::Outputs::ElasticSearch::TemplateManager.get_es_major_version elasticsearch 2.x detects major version is 2
     Failure/Error: allow(client).to receive(:get_version).and_return(JSON.parse(es_2x_version))

     NameError:
       uninitialized constant JSON
     # ./spec/unit/outputs/elasticsearch/template_manager_spec.rb:22:in `(root)'
     # ./vendor/bundle/jruby/1.9/gems/rspec-wait-0.0.9/lib/rspec/wait.rb:46:in `(root)'

  2) LogStash::Outputs::ElasticSearch::TemplateManager.get_es_major_version elasticsearch 5.x detects major version is 5
     Failure/Error: allow(client).to receive(:get_version).and_return(JSON.parse(es_5x_version))

     NameError:
       uninitialized constant JSON
     # ./spec/unit/outputs/elasticsearch/template_manager_spec.rb:30:in `(root)'
     # ./vendor/bundle/jruby/1.9/gems/rspec-wait-0.0.9/lib/rspec/wait.rb:46:in `(root)'

  3) LogStash::Outputs::ElasticSearch::TemplateManager.get_es_major_version elasticsearch 1.x detects major version is 1
     Failure/Error: allow(client).to receive(:get_version).and_return(JSON.parse(es_1x_version))

     NameError:
       uninitialized constant JSON
     # ./spec/unit/outputs/elasticsearch/template_manager_spec.rb:14:in `(root)'
     # ./vendor/bundle/jruby/1.9/gems/rspec-wait-0.0.9/lib/rspec/wait.rb:46:in `(root)'

  4) LogStash::Outputs::ElasticSearch::HttpClient join_bulk_responses when items key is not available should be handled properly
     Failure/Error:
       JSON.parse ('[{
         "took": 4,
         "errors": false
       }]')

     NameError:
       uninitialized constant JSON
     # ./spec/unit/outputs/elasticsearch/http_client_spec.rb:152:in `bulk_response'
     # ./spec/unit/outputs/elasticsearch/http_client_spec.rb:158:in `(root)'
     # ./vendor/bundle/jruby/1.9/gems/rspec-wait-0.0.9/lib/rspec/wait.rb:46:in `(root)'

```

So I'm thinking we do need this dep, and it should be a development dep. Have I missed something?

andrewvc replied:
I think you're right, I think it is a development dependency.
Logstash core does reference the JSON gem, and we only `LogStash::Util::Json` in the runtime portion of the code, but I guess the tests use JSON directly.

andrewvc commented on spec/integration/outputs/create_spec.rb:
```diff
@@ -3,7 +3,7 @@
   require "logstash/outputs/elasticsearch"
 
-  def get_es_output(action, id)
+  def get_es_output(action, id, external_version)
```
Can we just default the new parameter to `nil` or `false` to keep the 2-arity version valid?

andrewvc commented on spec/integration/outputs/delete_spec.rb:
```diff
@@ -0,0 +1,61 @@
+  end
+
+  context "when delete only" do
+    it "should ignore non-monotonic external_version updates" do
```
We should probably log in this situation no? Can we add a check there in these specs?

vvcephei replied:
Logging certainly seems appropriate. What level do you think? I'm thinking that it may be too verbose at INFO level; DEBUG seems more appropriate.

andrewvc commented on spec/integration/outputs/delete_spec.rb:
```diff
@@ -0,0 +1,61 @@
+    it "should ignore non-monotonic external_version updates" do
+      id = "ev2"
+      subject = get_es_output({ 'document_id' => id, "external_version" => "99", "action" => "index" })
+      subject.register
```
Can you rejigger these specs to use `before` blocks a bit better? Can you use the `subject {}` declaration in rspec itself with a separate `let(:options) { {'document_id' => id, ...}` declaration? It's a lot cleaner than chucking it all in the `it` block.

vvcephei replied:
Ok. Honestly I've very little clue as to what I'm doing here ;) I'll do a little research and incorporate this feedback.

andrewvc replied:
I'd take a quick spin through: http://betterspecs.org/ . If you still have questions I'm glad to work with you to bring this PR to completion :) It's hard to know how familiar people are/are not with RSpec :)

vvcephei replied:
Thanks!

andrewvc commented on spec/integration/outputs/delete_spec.rb:
```diff
@@ -0,0 +1,61 @@
+      subject.register
+      subject.multi_receive([LogStash::Event.new("message" => "foo")])
+      r = @es.get(:index => 'logstash-delete', :type => 'logs', :id => id, :refresh => true)
+      insist { r["_version"] } == 99
```
There's no way you could have known this, but some of our specs use this crazy `insist` syntax which we're trying to move away from. Can you rewrite these to be standard rspec. This would be `expect(r['_version']).to eq(99)`

andrewvc replied:
Also, it might make sense to use a dynamic version and have the version# be part of the document like in the real world rather than defining two different outputs.

vvcephei replied:
Ok on the expect syntax.

I tried for a while to get the dynamic version thing working. It seems like whatever is responsible for extracting the dynamic properties isn't wired up in these tests? Or maybe I just kept getting the syntax wrong. I couldn't find any examples of other tests that use dynamic properties. Do you have any context here?

I'll give it another shot.

vvcephei replied:
allrighty, on the dynamic version bit... It looks like the variable substitution isn't happening.

I tried testing with this snippet in delete_spec.rb:

```
  context "when delete only" do
    it "should ignore non-monotonic external version updates" do
      id = "ev2"
      subject = get_es_output({ 'document_id' => id, "version" => "%{my_version}", "version_type" => "external", "action" => "index" })
      subject.register
      subject.multi_receive([LogStash::Event.new("message" => "foo", "my_version" => 99)])
      ...
```

I get this exception: 

```
Failures:

  1) Versioned delete when delete only should ignore non-monotonic external version updates
     Failure/Error: LogStash::Outputs::ElasticSearch.new(settings.merge!(options))

     LogStash::ConfigurationError:
       Something is wrong with your configuration.
     # ./vendor/bundle/jruby/1.9/gems/logstash-core-5.1.2.snapshot1-java/lib/logstash/config/mixin.rb:125:in `config_init'
     # ./vendor/bundle/jruby/1.9/gems/logstash-core-5.1.2.snapshot1-java/lib/logstash/outputs/base.rb:63:in `initialize'
     # ./spec/integration/outputs/delete_spec.rb:15:in `get_es_output'
     # ./spec/integration/outputs/delete_spec.rb:31:in `(root)'
     # ./vendor/bundle/jruby/1.9/gems/rspec-wait-0.0.9/lib/rspec/wait.rb:46:in `(root)'
```

So I put a "puts" in mixin.rb to see what's up:

```
    if !self.class.validate(params)
      puts "params[#{params}]"
      raise LogStash::ConfigurationError,
        I18n.t("logstash.runner.configuration.invalid_plugin_settings")
    end
```

And this is what it spits out:

```
Versioned delete
  when delete only
params[{"manage_template"=>true, "index"=>"logstash-delete", "template_overwrite"=>true, "hosts"=>[//127.0.0.1:9200], "document_id"=>"ev2", "version"=>"%{my_version}", "version_type"=>"external", "action"=>"index", "id"=>"elasticsearch_647b6be3-dddd-4bce-8464-9ee0d84f4463", "enable_metric"=>true, "type"=>"", "tags"=>[], "exclude_tags"=>[], "codec"=><LogStash::Codecs::Plain id=>"plain_9bc0c20a-233c-4b26-b5b2-b669d1177838", enable_metric=>true, charset=>"UTF-8">, "workers"=>1, "template_name"=>"logstash", "parent"=>nil, "idle_flush_time"=>1, "upsert"=>"", "doc_as_upsert"=>false, "script"=>"", "script_type"=>"inline", "script_lang"=>"painless", "script_var_name"=>"event", "scripted_upsert"=>false, "retry_initial_interval"=>2, "retry_max_interval"=>64, "retry_on_conflict"=>1, "pipeline"=>nil, "ssl_certificate_verification"=>true, "sniffing"=>false, "sniffing_delay"=>5, "timeout"=>60, "failure_type_logging_whitelist"=>[], "pool_max"=>1000, "pool_max_per_route"=>100, "healthcheck_path"=>"/", "resurrect_delay"=>5, "validate_after_inactivity"=>10000}]
    should ignore non-monotonic external version updates (FAILED - 1)
    should commit monotonic external version updates

Failures:

  1) Versioned delete when delete only should ignore non-monotonic external version updates
     Failure/Error: LogStash::Outputs::ElasticSearch.new(settings.merge!(options))

     LogStash::ConfigurationError:
       Something is wrong with your configuration.
     # ./vendor/bundle/jruby/1.9/gems/logstash-core-5.1.2.snapshot1-java/lib/logstash/config/mixin.rb:126:in `config_init'
     # ./vendor/bundle/jruby/1.9/gems/logstash-core-5.1.2.snapshot1-java/lib/logstash/outputs/base.rb:63:in `initialize'
     # ./spec/integration/outputs/delete_spec.rb:15:in `get_es_output'
     # ./spec/integration/outputs/delete_spec.rb:31:in `(root)'
     # ./vendor/bundle/jruby/1.9/gems/rspec-wait-0.0.9/lib/rspec/wait.rb:46:in `(root)'
```

So you can see it's trying to initalize the "version" property to "%{my_version}", which fails validation as a number.

What am I doing wrong here?

Thanks.

vvcephei replied:
Oh. Interesting. If I remove the `:validate => :number` or replace it with `:validate => :string`, it works just fine...

Is this expected? I guess I'll just leave it as `:string`...

vvcephei replied:
I guess it makes sense, since the validation is at configuration time, before you see any documents.

andrewvc commented on spec/integration/outputs/index_version_spec.rb:
```diff
@@ -0,0 +1,88 @@
+  context "when index only" do
+    it "should default to ES version" do
+      subject = get_es_output({ 'document_id' => "123" })
+      subject.register
```
Love that you wrote tests here! Same deal here WRT `before` blocks and better leveraging `let`.

vvcephei commented:
Thanks for your feedback, @andrewvc. I've responded where conversation is needed. Otherwise, I'll incorporate the changes and update the PR asap.

vvcephei commented:
Also, I've had more time to stew about it, and I'm going to switch to the strategy of specifying version and version_type separately. It seems the more correct and obvious thing to do.

andrewvc commented:
@vvcephei I think that makes sense

vvcephei commented:
Ok, @andrewvc, I believe I have addressed all your concerns.

I read through http://betterspecs.org/ and *think* I made idiomatic usage of the subject block.

I also pulled in the validation and documentation that @neumaics had in his PR (#536). Perhaps it would be appropriate to add him as an author on this commit.

Please let me know if you have further feedback, and thanks for your help!

vvcephei commented:
oh, I should mention that I wasn't totally sure what you meant by "use `before` blocks a bit better". I gathered from the betterspecs page that they shouldn't be used to assign variables, but establishing a blank slate for the test seemed to be an appropriate usage to me.

andrewvc commented on lib/logstash/outputs/elasticsearch/common.rb:
```diff
@@ -120,9 +127,13 @@ def submit(actions)
         status = action_props["status"]
         failure  = action_props["error"]
         action = actions[idx]
+	action_params = action[1]
```
Indentation off by 1

andrewvc commented on lib/logstash/outputs/elasticsearch/common.rb:
```diff
@@ -120,9 +127,13 @@ def submit(actions)
 
         if SUCCESS_CODES.include?(status)
           next
+	elsif CONFLICT_CODE == status && VERSION_TYPES_PERMITTING_CONFLICT.include?(action_params[:version_type])
```
Indentation

andrewvc commented:
@vvcephei this is looking pretty close to done. There are only a few changes that need to be made.

The tests still need some work, but rspec can take a while to master. What I'll do is fork your PR branch and submit a PR to your PR :) with some cleanups for the specs.

Thanks for the excellent work and responsiveness, this is really one of the higher quality patches we've received :).

andrewvc commented:
@vvcephei I've opened up a spec fixes branch for this PR here: https://github.com/vvcephei/logstash-output-elasticsearch/pull/1

vvcephei commented:
Thanks so much for your help, dude. I just looked at it, and, yeah, I'm
glad you just took over ;). That syntax is cool, but I did not expect to be
able to declare subject and let in that order.

I will pull it in and take care of the other indentation problems when I
get a chance.

And to your earlier comment... No problem! The easier I make it for you to
accept the PR, the better for me! But seriously, I really appreciate the
time you've put in to helping me.

On Fri, Dec 23, 2016, 15:00 Andrew Cholakian <notifications@github.com>
wrote:

> @vvcephei <https://github.com/vvcephei> I've opened up a spec fixes
> branch for this PR here: vvcephei#1
> <https://github.com/vvcephei/logstash-output-elasticsearch/pull/1>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/logstash-plugins/logstash-output-elasticsearch/pull/528#issuecomment-269045136>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAy1E8lgfj9Pec2T7FW6tkQjhibd_ZJGks5rLDZ-gaJpZM4LJQls>
> .
>

vvcephei commented:
Ok @andrewvc , I think I'm all good, except that now the CLA check says " Error when checking the CLA status".

I don't know if you can force it to recheck, since it was previously passing.

Thanks again!

andrewvc commented:
@vvcephei I don't see the changes from https://github.com/vvcephei/logstash-output-elasticsearch/pull/1/files reflected here. Have you pushed up yet?

vvcephei commented:
@andrewvc It is there, but I made some additional changes. There were tabs mixed with spaces (my fault, this was related to the other indentation problems), and the `get_es_output` method was unused, so I removed it.

So now
* https://github.com/logstash-plugins/logstash-output-elasticsearch/pull/528/files?diff=split#diff-d8279618c95db2c4318c3d1b74876d4cR19 lines up with https://github.com/vvcephei/logstash-output-elasticsearch/pull/1/files#diff-d8279618c95db2c4318c3d1b74876d4cR31 from your PR
* https://github.com/logstash-plugins/logstash-output-elasticsearch/pull/528/files?diff=split#diff-d8279618c95db2c4318c3d1b74876d4cR26 lines up with https://github.com/vvcephei/logstash-output-elasticsearch/pull/1/files#diff-d8279618c95db2c4318c3d1b74876d4cR38
* etc.

If you're not seeing what I'm seeing in the diff, there may be a problem with Github. If you rm my repo and re-clone it, you should see the changes for sure.

I'll keep an eye out for further messages.
-John

vvcephei commented:
oh, I meant to say that that latest push was me removing the unused method (I just noticed it).

andrewvc commented on spec/integration/outputs/index_version_spec.rb:
```diff
@@ -0,0 +1,100 @@
+describe "Versioned indexing", :integration => true, :version_greater_than_equal_to_2x => true do
+  require "logstash/outputs/elasticsearch"
+
+  before :each do
```
Can you change this section to be:

```ruby
let(:es) { get_client }

before(:each) do
  # Delete all templates first.
  # Clean ES of data before we start.
  es.indices.delete_template(:name => "*")
  # This can fail if there are no indexes, ignore failure.
  es.indices.delete(:index => "*") rescue nil
  es.indices.refresh
end 
```

you will also have to generally replace `@es` with `es`

vvcephei commented:
\o/

andrewvc commented on spec/integration/outputs/delete_spec.rb:
```diff
@@ -0,0 +1,62 @@
+  require "logstash/outputs/elasticsearch"
+
+  before :each do
+    @es = get_client
```
Can you make the changes I've requested in `index_version_spec` apply here as well?

andrewvc commented:
@vvcephei sorry for the premature approval, I was confused about the test changes (my bad), but there are just a couple little things before I can merge this :)

vvcephei commented:
(╯°□°)╯︵ ┻━┻

just kidding ;)

I'll take care of it right now.

vvcephei commented:
@andrewvc I converted `@es` into `let` expressions in both `index_version_spec` and `delete_spec`.

andrewvc commented:
Released as 6.2.0 . Thanks for the great work @vvcephei !
